rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -505,8 +488,13 @@ extern "C" rmw_ret_t rmw_subscription_set_listener_callback(
 {
   auto sub = static_cast<CddsSubscription *>(rmw_subscription->data);
 
+  // Set a NULL listener while we update the user callback data.
+  dds_listener_t * listener = nullptr;
+  dds_get_listener(sub->enth, listener);
+  dds_delete_listener(listener);
+  dds_set_listener(sub->enth, NULL);
```

Comment:
This one call suffices for removing the listeners, the preceding three lines have no effect (this applies for all occurrences).